Fix reading/writing of track numbers including total tracks

uklotzde opened the pull request:
https://bugs.launchpad.net/mixxx/+bug/1530203

uklotzde committed:
Improve writing of Ogg::XiphComment tags and purge empty fields

uklotzde committed:
Improve writing of APE tags and purge empty items

uklotzde committed:
Improve writing of ID3v2 tags and purge empty frames

uklotzde committed:
Improve writing of MP4 tags and purge empty atoms

uklotzde committed:
Read track number including total tracks

uklotzde committed:
Write track number including total tracks

uklotzde committed:
Encapsulate track numbers in new class (with unit tests)

uklotzde committed:
Display track numbers as strings in table views

uklotzde committed:
Fix plugin builds (once again)

rryan commented on src/library/columncache.cpp:
```diff
@@ -81,7 +81,7 @@ void ColumnCache::setColumns(const QStringList& columns) {
     m_columnSortByIndex.insert(m_columnIndexByEnum[COLUMN_LIBRARYTABLE_COMPOSER], sortNoCase);
     m_columnSortByIndex.insert(m_columnIndexByEnum[COLUMN_LIBRARYTABLE_GROUPING], sortNoCase);
-    m_columnSortByIndex.insert(m_columnIndexByEnum[COLUMN_LIBRARYTABLE_TRACKNUMBER], sortInt);
+    m_columnSortByIndex.insert(m_columnIndexByEnum[COLUMN_LIBRARYTABLE_TRACKNUMBER], sortNoCase);
```
Even though it's a string, we currently sort by integer in case they happen to parse successfully as integers -- because users will complain that "10" sorts after "1" (and as far as I can remember nobody has complained about missing "total tracks" since I've worked on Mixxx).

Now that this is a structured type, can we define a sorting on TrackNumbers that behaves in a sane way when a total number of tracks is present?

rryan replied:
Sorry, that made no sense since we sort with sqlite. Maybe we need two columns?

uklotzde replied:
Currently the string found in the tags is preserved as much as possible. The same is true if a user edits track numbers in Mixxx.

I would propose to add a hidden Integer column "tracknumber_current" where we store only the current track number as an Integer if it can be parsed successfully from the track number string. This column should only be used internally in database queries for sorting purposes and should not be visible to the user.

Later we might decide to completely drop the track number string and split it into 2 integer values "tracknumber_current" and "tracknumber_total".

Are the terms "current" and "total" reasonable? I didn't find a more appropriate alternative for "current".

uklotzde replied:
The new column "tracknumber_current" needs to be populated during database migration, otherwise sorting by track number does not work as expected. This might take some time for huge libraries.

In this case we should already add both columns at once even if they are only used internally and their values are hidden.

daschuer replied:
What happens if the sorting remains at sortInt = cast(%1 as integer), sqlite does probably the job for us. 

An other thought: I have seen track numbers from vinyl rips, labelled A1 ... B2  
http://www.discogs.com/Rolf-Und-Seine-Freunde-Rolfs-Schulweg-Hitparade/release/4633319

So a final solution would be to register a custom sorter. This can also deal with the disk number field.

uklotzde committed:
More tests for track numbers with leading '0's

Test that track numbers with leading '0's are not parsed as octal numbers.

daschuer commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -535,28 +541,44 @@ void writeID3v2UserTextIdentificationFrame(
+// Bitmask of optional tag fields that should NOT be written into
+// the common part of the tag. For future extension it is safer
+// to explicitly specify the exceptions instead of the common case!
+enum WriteTagMask {
```
Please explain a bit more
// this is used for ... in case

daschuer commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -564,44 +586,72 @@ void writeTrackMetadataIntoTag(
+    // NOTE(uklotzde): The numeric track number does not reflect the
+    // total number of tracks! Derived tags might be able to write the
+    // complete string from trackMetadata.getTrackNumber() into the
+    // corresponding field.
```
What is a derived tag?

daschuer commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -785,6 +846,46 @@ void readTrackMetadataFromXiphComment(TrackMetadata* pTrackMetadata,
+                TrackNumbers::fromString(trackNumberText));
+        QString trackTotalText;
+        if ((TrackNumbers::ParseResult::VALID == trackNumbersFromString.second) &&
+                !trackNumbersFromString.first.hasTotal()) {
```
I am not sure, but It looks like If a metadata contains "TOTALTRACKS" it should be the first choice. 
Probably related: http://forums.mp3tag.de/index.php?showtopic=16223

uklotzde replied:
Nope: https://wiki.xiph.org/Field_names

I will add a comment with this reference.

On 01/02/2016 11:22 AM, Daniel Schürmann wrote:
>
> In src/track/trackmetadatataglib.cpp 
> <https://github.com/mixxxdj/mixxx/pull/835#discussion_r48683902>:
>
> > @@ -785,6 +846,46 @@ void readTrackMetadataFromXiphComment(TrackMetadata* pTrackMetadata,
> >                  toQStringFirstNotEmpty(tag.fieldListMap()["GROUPING"]));
> >      }
> >
> > +    if (tag.fieldListMap().contains("TRACKNUMBER")) {
> > +        const QString trackNumberText(
> > +                toQStringFirstNotEmpty(tag.fieldListMap()["TRACKNUMBER"]));
> > +        // Try to parse both the track number and the optional total tracks
> > +        // from this field
> > +        const std::pair<TrackNumbers, TrackNumbers::ParseResult> trackNumbersFromString(
> > +                TrackNumbers::fromString(trackNumberText));
> > +        QString trackTotalText;
> > +        if ((TrackNumbers::ParseResult::VALID == trackNumbersFromString.second) &&
> > +                !trackNumbersFromString.first.hasTotal()) {
>
> I am not sure, but It looks like If a metadata contains "TOTALTRACKS" it 
> should be the first choice.
> Probably related: http://forums.mp3tag.de/index.php?showtopic=16223
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/835/files#r48683902>.
>

daschuer commented:
What is the use case for displaying the total track number? 
IMHO it is only cluttering the track number column if we have tracks with total number extensions and track without. 

On Linux I have not tool that allows me to edit the track number in a 1/12 format. So we are probably only talking about a mp3 shortcoming. 

The real underlying issue, is that we must not override (loose) the total track number if we write the meta-data back. I see two good solutions for it: Introduce a total tracks column, or read the total track number before write. 

I actually can see a use case for the CD number Information. Since you currently cannot play a two disks album in recorded order using Mixxx. since Track 1 of disk 2 is played before or after  Track 1 of disk 2. 

Sometimes it is solved by prefix the track number with the disk number like that 101 and 201.

daschuer commented:
By the way: I have not a single "1/2" notation in my track number column if I view my tracknumber column with a sqlite viewer. But I have mp3 track with that notation in its metadata.  
So the total track number is already lost if a user migrates from previous Mixxx version. 

Conclusion: We Need to read before write to preserve total tracks.

uklotzde commented:
Mixxx is currently very, very dumb when parsing the track number from file 
tags. This is why I lost all my total track numbers on files that have been 
written by Mixxx.

But we cannot force the user to reload all metadata from the files, because 
then you will loose all your track info edits done in Mixxx. And since file 
tags are currently not updated users will complain about outdated track 
infos. We cannot force them to reload from files and I don't know of any 
strategy that is safe for all use cases :(

On 01/02/2016 11:50 AM, Daniel Schürmann wrote:
>
> By the way: I have not a single "1/2" notation in my track number column 
> if I view my tracknumber column with a sqlite viewer. But I have mp3 track 
> with that notation in its metadata.
>
> So the total track number is already lost if a user migrates from previous 
> Mixxx version.
>
> Conclusion: We Need to read before write to preserve total tracks.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/835#issuecomment-168379548>.
>

uklotzde committed:
Improve code comments for writing file tags

uklotzde committed:
Improve reading/writing of XiphComment total tracks

uklotzde commented:
Maybe I have an idea including a migration strategy for solving this issue:

Add new column "tracknumbers" with the (invalid) default value "/"

"tracknumber":

  * remains in the DB schema for backward compatibility
  * still stores only the current track number
  * is used for sorting and as a column in the table views
  * gets its value from "tracknumbers"
  * modifies only the current track number of "tracknumbers" when set

"tracknumbers":

  * will be populated automatically if value is "/":
      o read current track number from Mixxx library
      o read total tracks from the file
  * is used in the track properties dialog for editing both the current and
    total track number at once
  * is used for reading/writing file tags

On 01/02/2016 11:40 AM, Daniel Schürmann wrote:
>
> What is the use case for displaying the total track number?
> IMHO it is only cluttering the track number column if we have tracks with 
> total number extensions and track without.
>
> On Linux I have not tool that allows me to edit the track number in a 1/12 
> format. So we are probably only talking about a mp3 shortcoming.
>
> The real underlying issue, is that we must not override (loose) the total 
> track number if we write the meta-data back. I see two good solutions for 
> it: Introduce a total tracks column, or read the total track number before 
> write.
>
> I actually can see a use case for the CD number Information. Since you 
> currently cannot play a two disks album in recorded order using Mixxx. 
> since Track 1 of disk 2 is played before or after Track 1 of disk 2.
>
> Sometimes it is solved by prefix the track number with the disk number 
> like that 101 and 201.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/835#issuecomment-168379258>.
>

daschuer commented:
Do you know a tool that allows you to free-text edit TRACKNUMBER?

I would prefer the way foobar and EasyTag have solved the problem. 
They treat the tracknumber part after / as TRACKTOTAL. 

This means if a user ties to write a TRACKNUMBER without a TRACKTOTAL value, Mixxx has to read the TRACKNUMBER field and replace only the the first part in front of the / if there is any. 

If we think, Mixxx should be able to display and edit the TRACKTOTAL field, we need an additional column only for that. This new column needs a vergin value to distinguish "no value" from value not read because schema upgrade.

uklotzde commented:
Splitting at the first "/" into TRACKNUMBER and TRACKTOTAL sounds good. We 
can use the separator "/" itself as an initial value for TRACKTOTAL to 
trigger reloading from file.

With these 2 text fields we don't loose any information that is stored in 
the file tags. Restoring the original value is straightforward when writing 
tags.

On 01/02/2016 01:23 PM, Daniel Schürmann wrote:
>
> Do you know a tool that allows you to free-text edit TRACKNUMBER?
>
> I would prefer the way foobar and EasyTag have solved the problem.
> They treat the tracknumber part after / as TRACKTOTAL.
>
> This means if a user ties to write a TRACKNUMBER without a TRACKTOTAL 
> value, Mixxx has to read the TRACKNUMBER field and replace only the the 
> first part in front of the / if there is any.
>
> If we think, Mixxx should be able to display and edit the TRACKTOTAL 
> field, we need an additional column only for that. This new column needs a 
> vergin value to distinguish "no value" from value not read because schema 
> upgrade.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/835#issuecomment-168385041>.
>

uklotzde committed:
Add field for "track total" to TrackMetadata and TrackInfoObject

uklotzde committed:
Delete redundant "inline" keywords

uklotzde committed:
Read/write track number and total separately

uklotzde committed:
Revert "Display track numbers as strings in table views"

This reverts commit 0c3552af8e24c3d95710b11d561b73707e9d915e.

uklotzde committed:
Add new column "tracktotal" to Mixxx library

uklotzde commented:
Adding a new (text) column "tracktotal" as suggested by @daschuer was by far the easiest and cleanest solution. Now we don't need work on the sort order and migration of existing Mixxx libraries can be performed behind the scenes without the need to rescan everything. Also the tag reading/writing code has become much simpler.

uklotzde committed:
Add warning if reloading track total from file tags fails

uklotzde committed:
Merge branch 'master' into TrackNumbers

uklotzde committed:
Fix a copy&paste error

uklotzde committed:
Reload missing track numbers from file tags

uklotzde committed:
Adjust warning message

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -1293,6 +1303,38 @@ TrackPointer TrackDAO::getTrackFromDB(TrackId trackId) const {
+        // Just in case: Use track total from the track number itself if
+        // it can be split into two parts and the second part os not empty.
+        const std::pair<QString, QString> splittedTrackNumber(
+                TrackNumbers::splitString(pTrack->getTrackNumber()));
```
I do not like std::pair because of its anonymous members. 
I would use output parameters instead.

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -1293,6 +1303,38 @@ TrackPointer TrackDAO::getTrackFromDB(TrackId trackId) const {
+                        << location;
+            }
+        } else {
+            // Initialize track total with the value stored in the Mixxx library
```
I can't get the point why we can reach this. 
We have the TrackNumber column which does not contain a "1/12" format and we have the new metadata code which also not produces a "1/12" format.

daschuer commented on src/track/tracknumbers.cpp:
```diff
@@ -0,0 +1,106 @@
+//static
+std::pair<QString, QString> TrackNumbers::splitString(
+        const QString& str,
+        const QString& separator) {
```
It looks like separator is always default, so I you can just remove the parameter.

daschuer commented on src/track/tracknumbers.cpp:
```diff
@@ -0,0 +1,106 @@
+QString TrackNumbers::joinStrings(
+        const QString& currentText,
+        const QString& totalText,
+        const QString& separator) {
```
It looks like separator is always default here as well.

daschuer commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -564,44 +590,80 @@ void writeTrackMetadataIntoTag(
+    // Taglib::Tag might be able to write the complete string from
+    // trackMetadata.getTrackNumber() into some field or might parse the
+    // two values from the string including the total number of tracks.
+    // In this case parsing the track number string here should be omitted.
```
The comment needs to be updated.

daschuer commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -564,44 +590,80 @@ void writeTrackMetadataIntoTag(
+    // In this case parsing the track number string here should be omitted.
+    if (0 == (writeMask & WRITE_TAG_OMIT_TRACK_NUMBER)) {
+        // Set the numeric track number if available
+        const std::pair<TrackNumbers, TrackNumbers::ParseResult> trackNumbersFromString(
```
std::pair :-(

daschuer commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -785,6 +862,22 @@ void readTrackMetadataFromXiphComment(TrackMetadata* pTrackMetadata,
+    if (tag.fieldListMap().contains("TRACKNUMBER")) {
+        std::pair<QString, QString> splitted(
+                TrackNumbers::splitString(toQStringFirstNotEmpty(tag.fieldListMap()["TRACKNUMBER"])));
+        if (splitted.second.isEmpty()) {
```
I would remove this condition, and just override "second" if TRACKTOTAL or TOTALTRACKS is set. 
Since a "1/12" format is not mentioned in the Xiph specs.

daschuer commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -919,23 +1021,23 @@ bool writeTrackMetadataIntoID3v2Tag(TagLib::ID3v2::Tag* pTag,
+    writeID3v2TextIdentificationFrame(pTag, "TRCK",
+            TrackNumbers::joinStrings(
+                    trackMetadata.getTrackNumber(),
+                    trackMetadata.getTrackTotal()));
```
How is guaranteed that trackMetadata.getTrackTotal() has the value that is already in the metadata?

daschuer replied:
Nevermind, it is guaranteed since we now always read the value from DB or track.

daschuer commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -1058,12 +1184,36 @@ bool writeTrackMetadataIntoMP4Tag(TagLib::MP4::Tag* pTag, const TrackMetadata& t
+        break;
+    case TrackNumbers::ParseResult::VALID:
+        pTag->itemListMap()["trkn"] = TagLib::MP4::Item(
+                trackNumbersFromString.first.getCurrent(),
```
first??

uklotzde committed:
Remove unused function parameters

uklotzde committed:
Transform std::pair result into out parameters

uklotzde committed:
Remove unneeded data migration code from TrackDAO

uklotzde committed:
Update outdated comment about writing track numbers

uklotzde committed:
Read TRACKNUMBER/TRACKTOTAL according to the Xiph specs

uklotzde committed:
Add tests for splitting/joining track number strings

uklotzde committed:
Merge branch 'master' into TrackNumbers

uklotzde committed:
Renaming of track numbers: 'current' -> 'actual' (vs. 'total')

I finally found the term that I was looking for the whole time :)

uklotzde commented:
All review comments addressed

I've also added new unit tests for splitting/joining track number strings. The existing unit tests have proven to be very useful. They detected a copy&paste error during refactoring.

uklotzde committed:
Merge branch 'master' into TrackNumbers

daschuer commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -1010,44 +1146,45 @@ bool writeTrackMetadataIntoXiphComment(TagLib::Ogg::XiphComment* pTag,
+    // Remove the secondary/alternative field for total tracks to prevent
+    // inconsistencies if the tags are edited again by another external
+    // application that is unaware of "TRACKTOTAL".
+    pTag->removeField("TOTALTRACKS");
```
I a third party program uses only "TOTALTRACKS", it cannot access this info anymore after Mixxx touches the file.

uklotzde replied:
Potential inconsistency if Mixxx writes both "TRACKTOTAL" and "TOTALTRACKS":
1. External program reads "TOTALTRACKS" as written by Mixxx and ignores 
"TRACKTOTAL"
2. External program updates only "TOTALTRACKS" and leaves "TRACKTOTAL" untouched
3. Upon reloading Mixxx will ignore the new value of "TOTALTRACKS" and reads 
only the original value from "TRACKTOTAL"

This is what my comment is about and why I decided to remove the unofficial 
field "TOTALTRACKS" entirely when writing tags. If you don't see the value 
of that field in an external program you are not able to update it ;) Of 
course, this doesn't save you from re-creating it.

On 01/04/2016 08:27 AM, Daniel Schürmann wrote:
>
> In src/track/trackmetadatataglib.cpp 
> <https://github.com/mixxxdj/mixxx/pull/835#discussion_r48708686>:
>
> >
> > -    pTag->removeField("ALBUMARTIST");
> > -    pTag->addField("ALBUMARTIST",
> > -            toTagLibString(trackMetadata.getAlbumArtist()));
> > +    // According to https://wiki.xiph.org/Field_names "TRACKTOTAL" is
> > +    // the proposed field name, but some applications use "TOTALTRACKS".
> > +    writeXiphCommentField(pTag, "TRACKNUMBER",
> > +            toTagLibString(trackMetadata.getTrackNumber()));
> > +    writeXiphCommentField(pTag, "TRACKTOTAL",
> > +            toTagLibString(trackMetadata.getTrackTotal()));
> > +    // Remove the secondary/alternative field for total tracks to prevent
> > +    // inconsistencies if the tags are edited again by another external
> > +    // application that is unaware of "TRACKTOTAL".
> > +    pTag->removeField("TOTALTRACKS");
>
> I a third party program uses only "TOTALTRACKS", it cannot access this 
> info anymore after Mixxx touches the file.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/835/files#r48708686>.
>

uklotzde replied:
Maybe we should simply write both fields and accept the inconsistencies. 
This strategy at least preserves all data.

On 01/04/2016 10:50 AM, Uwe Klotz wrote:
> Potential inconsistency if Mixxx writes both "TRACKTOTAL" and "TOTALTRACKS":
> 1. External program reads "TOTALTRACKS" as written by Mixxx and ignores 
> "TRACKTOTAL"
> 2. External program updates only "TOTALTRACKS" and leaves "TRACKTOTAL" 
> untouched
> 3. Upon reloading Mixxx will ignore the new value of "TOTALTRACKS" and 
> reads only the original value from "TRACKTOTAL"
>
> This is what my comment is about and why I decided to remove the 
> unofficial field "TOTALTRACKS" entirely when writing tags. If you don't 
> see the value of that field in an external program you are not able to 
> update it ;) Of course, this doesn't save you from re-creating it.
>
> On 01/04/2016 08:27 AM, Daniel Schürmann wrote:
>>
>> In src/track/trackmetadatataglib.cpp 
>> <https://github.com/mixxxdj/mixxx/pull/835#discussion_r48708686>:
>>
>> >
>> > -    pTag->removeField("ALBUMARTIST");
>> > -    pTag->addField("ALBUMARTIST",
>> > -            toTagLibString(trackMetadata.getAlbumArtist()));
>> > +    // According tohttps://wiki.xiph.org/Field_names  "TRACKTOTAL" is
>> > +    // the proposed field name, but some applications use "TOTALTRACKS".
>> > +    writeXiphCommentField(pTag, "TRACKNUMBER",
>> > +            toTagLibString(trackMetadata.getTrackNumber()));
>> > +    writeXiphCommentField(pTag, "TRACKTOTAL",
>> > +            toTagLibString(trackMetadata.getTrackTotal()));
>> > +    // Remove the secondary/alternative field for total tracks to prevent
>> > +    // inconsistencies if the tags are edited again by another external
>> > +    // application that is unaware of "TRACKTOTAL".
>> > +    pTag->removeField("TOTALTRACKS");
>>
>> I a third party program uses only "TOTALTRACKS", it cannot access this 
>> info anymore after Mixxx touches the file.
>>
>> —
>> Reply to this email directly or view it on GitHub 
>> <https://github.com/mixxxdj/mixxx/pull/835/files#r48708686>.
>>
>

daschuer replied:
I just want to prevent a future bug "Mixxx deletes all my total track infos" which happens if an external App uses only the "TOTALTRACKS" tag. 
I would accept a future inconsistency if the user edits the total tracks in Mixxx, since it is not our responsibility that two fields exists and there is a user uses an "old" app which uses the wrong one. 

Changing the total track number is anyway a very rare action, but the current delete solution also kicks in if a user edits an unrelated field in Mixxx. 

How about update "TOTALTRACKS" if it exists or only delete "TOTALTRACKS" if the info dos not match "TRACKTOTAL".

uklotzde replied:
Good point! We should only write TOTALTRACKS if it already exists and 
otherwise ignore it.

On 01/04/2016 11:19 AM, Daniel Schürmann wrote:
>
> In src/track/trackmetadatataglib.cpp 
> <https://github.com/mixxxdj/mixxx/pull/835#discussion_r48717883>:
>
> >
> > -    pTag->removeField("ALBUMARTIST");
> > -    pTag->addField("ALBUMARTIST",
> > -            toTagLibString(trackMetadata.getAlbumArtist()));
> > +    // According to https://wiki.xiph.org/Field_names "TRACKTOTAL" is
> > +    // the proposed field name, but some applications use "TOTALTRACKS".
> > +    writeXiphCommentField(pTag, "TRACKNUMBER",
> > +            toTagLibString(trackMetadata.getTrackNumber()));
> > +    writeXiphCommentField(pTag, "TRACKTOTAL",
> > +            toTagLibString(trackMetadata.getTrackTotal()));
> > +    // Remove the secondary/alternative field for total tracks to prevent
> > +    // inconsistencies if the tags are edited again by another external
> > +    // application that is unaware of "TRACKTOTAL".
> > +    pTag->removeField("TOTALTRACKS");
>
> I just want to prevent a future bug "Mixxx deletes all my total track 
> infos" which happens if an external App uses only the "TOTALTRACKS" tag.
> I would accept a future inconsistency if the user edits the total tracks 
> in Mixxx, since it is not our responsibility that two fields exists and 
> there is a user uses an "old" app which uses the wrong one.
>
> Changing the total track number is anyway a very rare action, but the 
> current delete solution also kicks in if a user edits an unrelated field 
> in Mixxx.
>
> How about update "TOTALTRACKS" if it exists or only delete "TOTALTRACKS" 
> if the info dos not match "TRACKTOTAL".
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/835/files#r48717883>.
>

daschuer replied:
> Maybe we should simply write both fields and accept the inconsistencies. This strategy at least preserves all data.

Ok, out Should Mixxx introduce a "TOTALTRACKS"  field if there is non?

daschuer commented on src/test/tracknumberstest.cpp:
```diff
@@ -0,0 +1,137 @@
+    parseFromStringInvalid("-2 / 1", TrackNumbers(-2, 1));
+    parseFromStringInvalid("-3/-4", TrackNumbers(-3, -4));
+    parseFromStringInvalid("1/a", TrackNumbers(1, TrackNumbers::kValueUndefined));
+    parseFromStringInvalid("a /1", TrackNumbers(TrackNumbers::kValueUndefined, 1));
```
What should happen if a number is followed by rubbish: 
"2 Tracks" or " "1 of 12"  I have not seen this in the wild, but we should clarify by a test if we read it or not.

uklotzde replied:
Thanks, Daniel. After adding more test cases I found another minor 
inconsistency in the parsing algorithm ;)

On 01/04/2016 08:35 AM, Daniel Schürmann wrote:
>
> In src/test/tracknumberstest.cpp 
> <https://github.com/mixxxdj/mixxx/pull/835#discussion_r48708987>:
>
> > +TEST_F(TrackNumbersTest, ParseFromStringValid) {
> > +    parseFromStringValid("0", TrackNumbers(0));
> > +    parseFromStringValid("0/0", TrackNumbers(0, 0));
> > +    parseFromStringValid("1 / 0", TrackNumbers(1, 0));
> > +    parseFromStringValid(" 1\t/\t2  ", TrackNumbers(1, 2));
> > +    parseFromStringValid("12/7", TrackNumbers(12, 7));
> > +    parseFromStringValid("01234/12345", TrackNumbers(1234, 12345));
> > +}
> > +
> > +TEST_F(TrackNumbersTest, ParseFromStringInvalid) {
> > +    parseFromStringInvalid("-2", TrackNumbers(-2));
> > +    parseFromStringInvalid("1/ -1", TrackNumbers(1, -1));
> > +    parseFromStringInvalid("-2 / 1", TrackNumbers(-2, 1));
> > +    parseFromStringInvalid("-3/-4", TrackNumbers(-3, -4));
> > +    parseFromStringInvalid("1/a", TrackNumbers(1, TrackNumbers::kValueUndefined));
> > +    parseFromStringInvalid("a /1", TrackNumbers(TrackNumbers::kValueUndefined, 1));
>
> What should happen if a number is followed by rubbish:
> "2 Tracks" or " "1 of 12" I have not seen this in the wild, but we should 
> clarify by a test if we read it or not.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/835/files#r48708987>.
>

uklotzde committed:
Reuse splitting functionality for parsing of track numbers

uklotzde committed:
Add unit tests for parsing of invalid track numbers

uklotzde committed:
Provide public function for parsing of track number values

uklotzde committed:
Improve documentation of TrackNumbers class

uklotzde committed:
Pass parameter by value to prevent pointer aliasing

uklotzde committed:
Simplify and improve reading of XiphComment tags

uklotzde committed:
Align writing of XiphComment tags with reading

uklotzde committed:
Avoid Ogg::XiphComment::contains()

...and revert back to the previous check for existence of fields. This
is consistent with the other tag types.

daschuer commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -564,44 +590,104 @@ void writeTrackMetadataIntoTag(
+        const TagLib::Ogg::XiphComment& tag,
+        const TagLib::String& key,
+        QString* pValue = nullptr) {
+    if (tag.fieldListMap().contains(key)) {
```
Ah, I see, you version fixes the implicit insert of an empty StringList. 
But now we have again a double map lookup and the .isEmpty() is gone. 

How about: 
```
FieldListMap::ConstIterator it = tag.fieldListMap().find(key);
if (it != tag.fieldListMap().end() && !it.second.isEmpty()) {
    if (nullptr != pValue) {
        *pValue = toQStringFirstNotEmpty(it.second);
...
```
Unfortunately "it.second" relies in the DO_NOT_DOCUMENT part of the taglib header. 
Any other ideas?

uklotzde replied:
1) The double lookup is inevitable, because the map returned by 
fieldListMap() is immutable. It is only mutable for TagLib::MP4, but I 
consider this a flaw of the TagLib API. This is also the reason why I need 
to introduce the function getItemListMap(TagLib::MP4) with a const cast some 
time ago.

2) We currently do not interpret an empty string list as a missing tag for 
any tag type. If we want to handle it this way we have should introduce some 
helper functions for all tag types. I will change this, explicit functions 
for testing if a tag actually exists are needed anyway.

On 01/05/2016 07:53 PM, Daniel Schürmann wrote:
>
> In src/track/trackmetadatataglib.cpp 
> <https://github.com/mixxxdj/mixxx/pull/835#discussion_r48880846>:
>
> >      }
> >  }
> >
> > -template<typename T>
> > -inline void writeMP4Atom(TagLib::MP4::Tag* pTag, const TagLib::String& key,
> > -        const T& value) {
> > -    pTag->itemListMap()[key] = value;
> > +bool readXiphCommentField(
> > +        const TagLib::Ogg::XiphComment& tag,
> > +        const TagLib::String& key,
> > +        QString* pValue = nullptr) {
> > +    if (tag.fieldListMap().contains(key)) {
>
> Ah, I see, you version fixes the implicit insert of an empty StringList.
> But now we have again a double map lookup and the .isEmpty() is gone.
>
> How about:
>
> |FieldListMap::ConstIterator it = tag.fieldListMap().find(key); if (it != 
> tag.fieldListMap().end() && !it.second.isEmpty()) { if (nullptr != pValue) 
> { *pValue = toQStringFirstNotEmpty(it.second); ... |
>
> Unfortunately "it.second" relies in the DO_NOT_DOCUMENT part of the taglib 
> header.
> Any other ideas?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/835/files#r48880846>.
>

uklotzde replied:
Ups, my fault. Indeed, there is a double lookup that can be optimized ;)

On 01/05/2016 10:13 PM, Uwe Klotz wrote:
> 1) The double lookup is inevitable, because the map returned by 
> fieldListMap() is immutable. It is only mutable for TagLib::MP4, but I 
> consider this a flaw of the TagLib API. This is also the reason why I need 
> to introduce the function getItemListMap(TagLib::MP4) with a const cast 
> some time ago.
>
> 2) We currently do not interpret an empty string list as a missing tag for 
> any tag type. If we want to handle it this way we have should introduce 
> some helper functions for all tag types. I will change this, explicit 
> functions for testing if a tag actually exists are needed anyway.
>
> On 01/05/2016 07:53 PM, Daniel Schürmann wrote:
>>
>> In src/track/trackmetadatataglib.cpp 
>> <https://github.com/mixxxdj/mixxx/pull/835#discussion_r48880846>:
>>
>> >      }
>> >  }
>> >
>> > -template<typename T>
>> > -inline void writeMP4Atom(TagLib::MP4::Tag* pTag, const TagLib::String& key,
>> > -        const T& value) {
>> > -    pTag->itemListMap()[key] = value;
>> > +bool readXiphCommentField(
>> > +        const TagLib::Ogg::XiphComment& tag,
>> > +        const TagLib::String& key,
>> > +        QString* pValue = nullptr) {
>> > +    if (tag.fieldListMap().contains(key)) {
>>
>> Ah, I see, you version fixes the implicit insert of an empty StringList.
>> But now we have again a double map lookup and the .isEmpty() is gone.
>>
>> How about:
>>
>> |FieldListMap::ConstIterator it = tag.fieldListMap().find(key); if (it != 
>> tag.fieldListMap().end() && !it.second.isEmpty()) { if (nullptr != 
>> pValue) { *pValue = toQStringFirstNotEmpty(it.second); ... |
>>
>> Unfortunately "it.second" relies in the DO_NOT_DOCUMENT part of the 
>> taglib header.
>> Any other ideas?
>>
>> —
>> Reply to this email directly or view it on GitHub 
>> <https://github.com/mixxxdj/mixxx/pull/835/files#r48880846>.
>>
>

daschuer replied:
I do not understand 1) please give an example. 
Are you afraid the "it" is getting invalid in my code snipped? How can this happen? 
fieldListMap() just returns d->fieldListMap. If we cannot use d->fieldListMap twice, some internal taglib code will fail as well.

uklotzde replied:
I mixed up reading/writing when I quickly wrote the answer. The double 
lookup is in the read function can be eliminated as you suggested :)

I will eliminate most of the double lookups by encapsulating reading/writing 
of key/value pairs from tags in separate functions. At least for the common 
cases where the type of value is a string. Double lookups for some special 
cases will remain, because it would complicate the code too much.

On 01/05/2016 10:29 PM, Daniel Schürmann wrote:
>
> In src/track/trackmetadatataglib.cpp 
> <https://github.com/mixxxdj/mixxx/pull/835#discussion_r48898646>:
>
> >      }
> >  }
> >
> > -template<typename T>
> > -inline void writeMP4Atom(TagLib::MP4::Tag* pTag, const TagLib::String& key,
> > -        const T& value) {
> > -    pTag->itemListMap()[key] = value;
> > +bool readXiphCommentField(
> > +        const TagLib::Ogg::XiphComment& tag,
> > +        const TagLib::String& key,
> > +        QString* pValue = nullptr) {
> > +    if (tag.fieldListMap().contains(key)) {
>
> I do not understand 1) please give an example.
> Are you afraid the "it" is getting invalid in my code snipped? How can 
> this happen?
> fieldListMap() just returns d->fieldListMap. If we cannot use 
> d->fieldListMap twice, some internal taglib code will fail as well.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/835/files#r48898646>.
>

uklotzde committed:
Unify and optimize tag reading/writing

Eliminate various double lookups and unify handling of
StringList values as far as possible. The remaining double
lookups don't need to be removed.

uklotzde committed:
Merge branch 'master' into TrackNumbers

uklotzde committed:
Merge branch 'master' into TrackNumbers

uklotzde committed:
Merge branch 'master' into TrackNumbers

daschuer commented:
Cool B-) 
I did not notice that there where so much room for speedup. 
LGTM.
Thank you.
